### PR TITLE
feat(security): harden repository scan integrity

### PR DESCRIPTION
**Pull request by sozercan**
## Summary

- add run-scoped provenance and immutable receipts for repository targets, stages, findings, assessments, task inputs, and sealed bundles
- make output ingestion, validation, patching, and history projection fail closed across retries, task recreation, stale writers, and repository generations
- add accountable mapper inventories, strict validation artifacts, deterministic identities, quality metrics, and bounded canonical bundle verification
- expose quality, coverage, bundle, and history state through the API, CLI, and UI, with regenerated CRDs, staging manifests, and command documentation

## Why

Repository scan results crossed several mutable boundaries where task names, late writers, stale status, or incomplete inventory accounting could override authoritative state. This change binds accepted records to an exact repository incarnation, run, target, producer, and execution identity while retaining compatibility modes as explicitly unverified.

Advanced enforcement modes remain unavailable until their lease, capability-snapshot, and authorization prerequisites are present, so unsupported configurations continue to fail closed.

## Validation

- `make lint-fix`
- `make test`
- `cd ui && bun run lint && bun run test`
- `make build`
- focused race tests for SQLite stores, controller scan paths, API scan paths, workers, tools, security packages, and CLI
- `make generate && make manifests`
- `make docs-cli-check`
- Helm lint plus default and explicit-null renders for source and staging charts
- `bash -n scripts/*.sh`
- `git diff --check`

Root release snapshots under `deploy/` and `charts/orka/` remain unchanged.

### CODE REVIEW

File: internal/api/security_handlers.go [hunk 1]
```diff
@@ -972,77 +2020,526 @@
...
+		return err
+	}
+	pagination, err := parseFindingHistoryPagination(c)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, err.Error())
+	}
+	finding := authorized.finding
+	runAuthorization := make(map[string]bool)
+	items, next, err := listAuthorizedFindingHistory(
+		int(pagination.Limit),
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go]: Fixed in 7061c93. Finding-history pagination now parses and bounds the limit as an `int` before it reaches the store, so this handler no longer narrows an `int64`. Added boundary, malformed, and overflow coverage; `make lint-fix && make test` pass.

File: internal/api/security_handlers.go [hunk 2]
```diff
@@ -972,77 +2020,526 @@
...
-	finding, err := h.securityStore.GetFinding(c.Context(), namespace, c.Params("id"))
+	pagination, err := parseFindingHistoryPagination(c)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, err.Error())
+	}
+	finding := authorized.finding
+	runAuthorization := make(map[string]bool)
+	occurrences := make(map[string]*store.FindingOccurrence)
+	items, next, err := listAuthorizedFindingHistory(
+		int(pagination.Limit),
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go]: Fixed in 7061c93 with the shared bounded `int` pagination path used by all three finding-history handlers. The narrowing conversion is removed and pagination edge cases are covered by tests.

File: internal/api/security_handlers.go [hunk 3]
```diff
@@ -972,77 +2020,526 @@
...
+		return err
+	}
+	pagination, err := parseFindingHistoryPagination(c)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, err.Error())
+	}
+	finding := authorized.finding
+	runAuthorization := make(map[string]bool)
+	items, next, err := listAuthorizedFindingHistory(
+		int(pagination.Limit),
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go]: Fixed in 7061c93. Assessments now consume the same explicitly bounded `int` limit as occurrences and decisions, eliminating the narrowing conversion. Full Go tests pass.

File: internal/api/security_handlers.go [hunk 4]
```diff
@@ -238,66 +461,488 @@ func (h *Handlers) createSecurityScanRun(ctx context.Context, ui *UserInfo, scan
...
+) error {
+	if _, err := h.findActiveSecurityScanRunByIdempotencyKey(ctx, scan, requestKey); err == nil {
+		return nil
+	} else if !errors.Is(err, store.ErrNotFound) {
+		return securityIntegrityHTTPError(err, "find active scan run")
+	}
+
+	active, err := h.hasActiveSecurityScanPipelineTask(ctx, scan)
+	if err != nil || !active {
+		return err
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go:563]: Fixed in 7061c93. Active scan ownership is now enforced atomically by a partial unique index on `(namespace, repository_scan)`, independent of request key. API/controller contention paths and reservation release are covered by focused tests.

File: ui/src/components/security/repository-list.tsx [hunk 5]
```diff
@@ -53,56 +430,364 @@ export function RepositoryList() {
...
-  const lastScanAt = repo.status?.lastScanAt ?? repo.status?.lastSuccessfulScanAt
+function RepositoryCard({
+  repo,
+  repositoryQueryBinding,
+}: {
+  repo: RepositoryScan;
+  repositoryQueryBinding: LatestRunBinding;
+}) {
+  const runScan = useRunSecurityScan(repo.metadata.name);
+  const scanRuns = useScanRuns(repo.metadata.name);
```
[the remark this answers is not in the document]
Comment (sozercan) [reply ui/src/components/security/repository-list.tsx]: Fixed in 7061c93. The repository list requests `includeLatestRuns=true`; the API performs one batched SQLite latest-run query keyed by repository name/UID/generation, and cards consume the parent result instead of polling individually. Store, API, hook, and component tests were added.

File: internal/api/security_handlers.go [hunk 6]
```diff
@@ -238,66 +461,488 @@ func (h *Handlers) createSecurityScanRun(ctx context.Context, ui *UserInfo, scan
...
-				labels.LabelSecurityStage:     security.StageValidation,
-				labels.LabelSecurityFindingID: finding.ID,
+				labels.LabelManaged:              "true",
+				labels.LabelCreatedBy:            "repository-security",
+				labels.LabelSecurityTarget:       labels.SelectorValue(scan.Name),
+				labels.LabelSecurityScanID:       finding.ScanRunID,
+				labels.LabelSecurityMode:         security.StageValidation,
+				labels.LabelSecurityStage:        security.StageValidation,
+				labels.LabelSecurityFindingID:    finding.ID,
+				labels.LabelSecurityOccurrenceID: finding.CurrentOccurrenceID,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go]: Fixed in 7061c93. Task labels now use deterministic label-safe selector values while exact finding/occurrence IDs remain in environment bindings. Controller ingestion verifies the label/env pair and uses the full IDs; full-width and fail-closed cases are covered.

File: internal/tools/task_output_binding.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	requireBound := toolTaskRequiresBoundOutput(task)
+	if task.UID != "" {
+		if bound, ok := any(toolCtx.ResultStore).(store.BoundOutputStore); ok {
+			result, err := bound.GetBoundResult(ctx, task.Namespace, task.Name, string(task.UID), toolTaskOutputAttempt(task))
+			if err == nil {
+				return result.Data, nil
+			}
+			if requireBound {
+				return nil, err
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/tools/task_output_binding.go]: Fixed in 7061c93. Tool result reads now receive the configured binding mode: off uses legacy, audit prefers bound then falls back, and enforce remains bound-only/fail-closed. Propagation and fetch/wait behavior are covered across tools, API compatibility paths, and the harness broker.

File: internal/store/sqlite/security_integrity_store.go [hunk 8]
```diff
@@ -0,0 +1,1693 @@
...
+			evidence_receipt_ids_json TEXT NOT NULL DEFAULT '[]',
+			normalized_payload BLOB NOT NULL DEFAULT X'',
+			payload_digest TEXT NOT NULL DEFAULT '',
+			projection_validation_status TEXT NOT NULL DEFAULT '',
+			projection_evidence_json TEXT NOT NULL DEFAULT '[]',
+			record_digest TEXT NOT NULL,
+			created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
+		)`,
+		`CREATE TABLE IF NOT EXISTS security_finding_decisions (
+			id TEXT PRIMARY KEY,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/store/sqlite/security_integrity_store.go]: Fixed in 7061c93. Finding-decision identity is now `PRIMARY KEY(namespace, id)`, and transactional replay lookup includes namespace. Tests cover cross-namespace ID reuse, correct replay, and same-namespace duplicate mismatch.

File: internal/api/security_record_binding.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	run, err := h.securityStore.GetScanRun(ctx, scan.Namespace, runID)
+	if errors.Is(err, store.ErrNotFound) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return run.RepositoryScan == scan.Name &&
+		run.RepositoryScanUID != "" && run.RepositoryScanGeneration > 0 &&
+		run.RepositoryScanUID == string(scan.UID), nil
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_record_binding.go]: Fixed in c16308a. Run-bound record checks now require both the current RepositoryScan UID and generation. Tests cover stale-generation rejection, current-generation acceptance, and recreated-UID rejection; full Go tests pass.

File: internal/controller/repositoryscan_controller.go [hunk 10]
```diff
@@ -1385,1019 +1925,3214 @@ func threatModelLooksLikeToolTranscript(content string) bool {
...
+		return false, nil
+	}
+	var tasks corev1alpha1.TaskList
+	if err := r.List(ctx, &tasks,
+		client.InNamespace(scan.Namespace),
+		client.MatchingLabels(map[string]string{
 			labels.LabelSecurityTarget:    labels.SelectorValue(scan.Name),
-			labels.LabelSecurityFindingID: findingID,
+			labels.LabelSecurityFindingID: labels.SelectorValue(findingID),
 			labels.LabelSecurityStage:     security.StageValidation,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/controller/repositoryscan_controller.go:3756]: Fixed in b003c10. Active validation lookup now requires the exact finding ID, scan-run ID, and current occurrence ID from verified label/environment bindings. Prior-run and prior-occurrence tasks no longer suppress current scheduling; malformed current bindings remain fail-closed.

File: internal/api/security_handlers.go [hunk 11]
```diff
@@ -238,66 +461,488 @@ func (h *Handlers) createSecurityScanRun(ctx context.Context, ui *UserInfo, scan
...
+	}
+	if !activeSecurityScanRunPhase(run.Phase) || run.CompletedAt != nil {
+		return nil, fmt.Errorf("active scan run %q is no longer repairable", run.ID)
+	}
+	if run.Namespace != scan.Namespace || run.RepositoryScan != scan.Name ||
+		run.RepositoryScanUID != string(scan.UID) || run.RepositoryScanGeneration != scan.Generation {
+		return nil, fmt.Errorf("active scan run %q does not match the current repository scan identity", run.ID)
+	}
+	if run.Mode != mode || run.BaseCommit != baseCommit || run.HeadCommit != headCommit ||
+		run.ScannerPolicyVersion != security.ScannerPolicyVersion || run.PolicyDigest != policyDigest ||
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go]: Fixed in b003c10. Active-request recovery now uses the immutable request idempotency key and original request fields, so a mapper-resolved `HeadCommit` does not invalidate replay of the same in-flight request. Legacy rows without the immutable key retain strict head comparison.

File: internal/api/security_handlers.go [hunk 12]
```diff
@@ -972,77 +2051,530 @@ func (h *Handlers) GetSecurityFinding(c fiber.Ctx) error {
...
+	if finding.ScanRunID == "" {
+		return nil, fiber.NewError(fiber.StatusConflict, "legacy finding has no verifiable scan run")
+	}
+	run, runErr := h.securityStore.GetScanRun(c.Context(), namespace, finding.ScanRunID)
+	if runErr != nil && !errors.Is(runErr, store.ErrNotFound) {
+		return nil, fiber.NewError(fiber.StatusInternalServerError, fmt.Sprintf("failed to load finding scan run: %v", runErr))
+	}
+	if errors.Is(runErr, store.ErrNotFound) || run.RepositoryScanUID == "" {
+		if write {
+			return nil, fiber.NewError(fiber.StatusConflict, "legacy finding has no verifiable RepositoryScan UID")
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go:2311]: Kept this path fail-closed after deeper review. Pre-integrity rows have no durable RepositoryScan UID/generation or agent authorization receipt; using the current same-name object would allow a recreated or changed RepositoryScan to mutate a prior owner's finding. `TestLegacyFindingStateEndpointsRejectPreIntegrityUnboundRun` now documents the boundary. A future migration can enable this only after backfilling trusted ownership provenance.

File: internal/api/security_handlers.go [hunk 13]
```diff
@@ -279,94 +503,525 @@ func (h *Handlers) createSecurityScanRun(ctx context.Context, ui *UserInfo, scan
...
+		if occurrence.RepositoryScan != scan.Name || occurrence.ScanRunID != sourceRun.ID || occurrence.RunUID != sourceRun.RunUID ||
+			occurrence.PublicFindingID != finding.ID || occurrence.TargetReceiptID != sourceRun.TargetReceiptID ||
+			occurrence.TargetSHA != sourceRun.HeadCommit {
+			return fiber.NewError(fiber.StatusConflict, "finding occurrence does not match the source scan run")
+		}
+	}
+	if h.securityIntegrityStore != nil && strings.TrimSpace(finding.CurrentOccurrenceID) != "" {
+		assessments, _, err := h.securityIntegrityStore.ListFindingAssessments(ctx, store.FindingAssessmentFilter{
+			Namespace: scan.Namespace, RepositoryScan: finding.RepositoryScan, OccurrenceID: finding.CurrentOccurrenceID,
+			Kind: store.FindingAssessmentValidation, Limit: 1,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go]: Fixed in 980c5da6: the manual-validation admission check now lists up to 1000 validation assessments for the occurrence instead of only the oldest row, so any assessment satisfying validationAssessmentBlocksManualRequest blocks the rerun. Verified by internal/api tests (full package green).

File: internal/controller/repositoryscan_controller.go [hunk 14]
```diff
@@ -998,23 +1398,64 @@ func (r *RepositoryScanReconciler) progressLatestScanRun(ctx context.Context, sc
...
-	if !state.hasSucceededThreatModel {
-		return false, nil
-	}
-
 	run, err := r.SecurityStore.GetScanRun(ctx, scan.Namespace, scanID)
 	if err != nil {
 		return false, err
 	}
-	if run.Phase == scanRunPhaseSucceeded || run.Phase == scanRunPhaseFailed {
+	if run.Phase == scanRunPhaseFailed {
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/controller/repositoryscan_controller.go]: Fixed in 980c5da6: progressLatestScanRun now calls retireStaleScanRunReservation before progressing — an active run whose stored RepositoryScan UID/generation explicitly mismatches the live object is terminalized (markScanRunTerminalError) instead of launching mapper/review work with the new mutable spec, which also releases the active-repository reservation so a current-generation run can start. Regression test: TestProgressLatestScanRunRetiresStaleGenerationReservation.

File: internal/store/sqlite/security_integrity_store.go [hunk 15]
```diff
@@ -0,0 +1,1699 @@
...
+		return fmt.Errorf("migration failed: preflight active-repository index: %w", err)
+	}
+	if duplicateGroups != 0 {
+		return fmt.Errorf("migration failed: %d duplicate active repository groups remain after sealing reconciliation", duplicateGroups)
+	}
+
+	if _, err := tx.Exec(`CREATE UNIQUE INDEX idx_security_scan_runs_active_repository
+		ON security_scan_runs(namespace, repository_scan)
+		WHERE run_uid <> '' AND request_idempotency_key <> ''
+		  AND (phase IN ('pending', 'running') OR bundle_status = 'sealing')`); err != nil {
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/store/sqlite/security_integrity_store.go:464]: Fixed in 980c5da6: added the orka.ai/repositoryscan-run-reservation finalizer. On deletion, releaseScanRunReservations pages through the scan's runs and terminalizes every pending/running/sealing reservation (sealing bundle status transitions to failed) before the finalizer is removed, so the namespace+name active-run unique index cannot stay occupied by a deleted owner. Regression test: TestRepositoryScanDeletionReleasesActiveRunReservation, which also verifies a recreated RepositoryScan with a new UID can reserve its initial run.

File: internal/controller/repositoryscan_controller.go [hunk 16]
```diff
@@ -998,23 +1488,67 @@ func (r *RepositoryScanReconciler) progressLatestScanRun(ctx context.Context, sc
...
 		return false, nil
 	}
+	if retired, err := r.retireStaleScanRunReservation(ctx, scan, run); err != nil || retired {
+		return retired, err
+	}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/controller/repositoryscan_controller.go:1521]: Fixed in d4ee79a6: retireStaleScanRunReservation now explicitly clears status.lastScanID/lastScanTaskName and returns the CR to Pending (with a Ready=False/Pending condition) after terminalizing the mismatched run, since projection intentionally skips mismatched runs. The regression test asserts the cleared binding so a current-generation run can start.

File: internal/api/security_output_binding.go [hunk 17]
```diff
@@ -0,0 +1,341 @@
...
+	}
+	switch {
+	case filename == security.ArtifactSlices:
+		return stage == security.StageMapper
+	case filename == security.ArtifactThreatModel:
+		return stage == security.StageThreatModel
+	case filename == security.ArtifactFindingsV2:
+		return stage == security.StageReview
+	case strings.HasPrefix(filename, "security-review-context-") && strings.HasSuffix(filename, ".json"):
+		return stage == security.StageReview
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_output_binding.go]: Fixed in d4ee79a6: securityArtifactNameAllowedForStage now permits security-review-context-*.json for the mapper stage (its actual producer) in addition to review-stage re-uploads, so bound provenance is recorded instead of an audit-mode denial or enforce-mode rejection.

File: internal/controller/repositoryscan_controller.go [hunk 18]
```diff
@@ -1380,72 +2029,399 @@ func (r *RepositoryScanReconciler) persistThreatModelIfChanged(
...
+			TaskUID:  string(task.UID),
+			Attempt:  int64(task.Status.Attempts),
+			JobName:  task.Status.JobName,
+		},
+	}
+	if task.Annotations != nil {
+		runtimeSessionID := strings.TrimSpace(securityTaskRuntimeSessionID(task))
+		turnID := strings.TrimSpace(securityTaskTurnID(task))
+		correlationID := strings.TrimSpace(task.Annotations["orka.ai/harness-wrapper-correlation-id"])
+		if runtimeSessionID != "" && turnID != "" && correlationID != "" {
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/controller/repositoryscan_controller.go]: Fixed in d4ee79a6: the stage-receipt provenance selector no longer requires the removed wrapper correlation annotation. securityTaskExecutionCorrelationID derives a deterministic correlation from the controller-owned ACP attempt identity (acp:<runtimeSessionUID>:<attempt>:<promptID>), so built-in agent executions are recorded as harness provenance with runtime session and turn identity instead of Kubernetes executions with an empty Job name.

File: internal/controller/repositoryscan_controller.go [hunk 19]
```diff
@@ -151,6 +260,28 @@ func (r *RepositoryScanReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+			if err := r.Patch(ctx, scan, patch); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+	if r.SecurityStore != nil && !controllerutil.ContainsFinalizer(scan, repositoryScanRunFinalizer) {
+		patch := client.MergeFrom(scan.DeepCopy())
+		controllerutil.AddFinalizer(scan, repositoryScanRunFinalizer)
+		if err := r.Patch(ctx, scan, patch); err != nil {
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/controller/repositoryscan_controller.go:280]: Fixed in be091191: CreateRepositoryScan now stamps the orka.ai/repositoryscan-run-reservation finalizer on the object it creates, and createSecurityScanRun re-ensures the finalizer (rejecting deleting scans) before any durable run is reserved, so the pre-reconciliation delete window can no longer leak the reservation.

File: workers/general/main.go [hunk 20]
```diff
@@ -303,35 +310,416 @@ func runSecurityMapper(ctx context.Context) error {
...
 		ChangedFiles:         changedFiles,
 		ChangedLineRanges:    changedLineRanges,
 		DiffSummary:          diffSummary,
 		ChangedFilesError:    changedFilesError,
-		Slices:               slices,
+		CoverageReasonCodes:  coverageReasonCodes,
+		InventorySummary:     &inventorySummary,
+		DiscoveredFiles:      mapped.DiscoveredFiles,
+		ReviewableFiles:      mapped.ReviewableFiles,
+		OmittedFiles:         mapped.OmittedFiles,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply workers/general/main.go:421]: Fixed in be091191: marshalMapperArtifactWithinBudget keeps security-slices.json under a 9 MiB budget (below the uploader's 10 MiB skip threshold). When full inventories would exceed it, the detailed discovered/reviewable/omitted path lists are dropped — the InventorySummary counts and slices remain — with an explicit inventory_truncated coverage reason code; if even that exceeds the budget the mapper fails loudly instead of silently losing the authoritative artifact. Unit tests cover both the truncation and the small-inventory pass-through.

File: internal/api/security_handlers.go [hunk 21]
```diff
@@ -408,10 +1238,36 @@ func (h *Handlers) createSecurityPatchTask(ctx context.Context, ui *UserInfo, sc
...
 	agentRef := securityPatchAgentRef(scan)
 
-	taskName := security.PatchTaskName(scan.Name, finding.ID)
-	proposalID := security.PatchProposalID(taskName)
+	patchScopeID := strings.TrimSpace(finding.CurrentOccurrenceID)
+	if patchScopeID == "" {
+		patchScopeID = strings.TrimSpace(finding.ID)
+	}
+	taskName := security.ScanStageTaskNameForRun(
+		scan.Name, "patch", security.StagePatch, patchScopeID, sourceRun.RunUID,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go:1322]: Fixed in be091191: createSecurityPatchTask now records the durable proposal before creating the deterministic Task. On proposal-insert conflict it recovers the matching pending proposal for the same run/occurrence binding; on Task AlreadyExists it validates the existing Task against the deterministic binding via securityTaskMatchesExpected. A transient failure between the two writes is now retryable instead of stranding an orphan Task.

File: workers/general/main.go [hunk 22]
```diff
@@ -303,35 +310,449 @@ func runSecurityMapper(ctx context.Context) error {
...
+	data, err := json.MarshalIndent(artifact, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+	if len(data) <= maxMapperArtifactBytes {
+		return data, nil
+	}
+	artifact.OmittedFiles = nil
+	artifact.DiscoveredFiles = nil
+	artifact.ReviewableFiles = nil
```
[the remark this answers is not in the document]
Comment (sozercan) [reply workers/general/main.go:353]: Fixed in 03f3f79c: the budget helper is now schema-aware. Schema-v1 artifacts keep the degraded truncation (nil inventories are valid for v1 parsing). Pinned-target schema-v2 artifacts fail loudly with an explicit over-budget error instead of emitting an artifact that validateMapperInventories would reject — the accountable inventory cannot be truncated without breaking the contract, and a visible mapper failure is the fail-closed behavior pinned mode intends. Covered by TestMarshalMapperArtifactWithinBudgetFailsClosedForPinnedV2.

File: internal/api/security_handlers.go [hunk 23]
```diff
@@ -279,94 +506,543 @@ func (h *Handlers) createSecurityScanRun(ctx context.Context, ui *UserInfo, scan
...
+
+// ensureRepositoryScanRunFinalizer guarantees the run-reservation finalizer is
+// persisted before any durable scan run is reserved, closing the window where
+// deleting a freshly created RepositoryScan could leak its reservation.
+func (h *Handlers) ensureRepositoryScanRunFinalizer(ctx context.Context, scan *corev1alpha1.RepositoryScan) error {
+	if scan == nil || !scan.DeletionTimestamp.IsZero() {
+		return fiber.NewError(fiber.StatusConflict, "repository scan is being deleted")
+	}
+	if slices.Contains(scan.Finalizers, security.RepositoryScanRunFinalizer) {
+		return nil
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/api/security_handlers.go:756]: Fixed in 03f3f79c: terminalizeScanRunIfOwnerGone runs after CreateScanRunWithTaskInput (and before recovering an existing reservation). It re-reads the RepositoryScan and, when the owner is deleting, gone, or recreated with a different UID, terminalizes the just-created run (releasing the active-repository reservation) and returns 409. Since the controller's finalizer release lists runs before removing the finalizer, any reservation inserted after that listing is observed here with the deletion already in flight and self-releases.

File: internal/controller/repositoryscan_controller.go [hunk 24]
```diff
@@ -602,20 +1009,112 @@ func (r *RepositoryScanReconciler) markScanRunTerminalError(ctx context.Context,
...
+// (or the object was recreated) after the run was reserved, so the stale
+// reservation must never progress with the new mutable spec; retiring it lets
+// a current-generation run start.
+func (r *RepositoryScanReconciler) retireStaleScanRunReservation(
+	ctx context.Context,
+	scan *corev1alpha1.RepositoryScan,
+	run *store.ScanRun,
+) (bool, error) {
+	if !activeScanRunPhase(run.Phase) || !scanRunExplicitlyMismatchesRepositoryScan(run, scan) {
+		return false, nil
```
[the remark this answers is not in the document]
Comment (sozercan) [reply internal/controller/repositoryscan_controller.go]: Fixed in 03f3f79c: retireStaleScanRunReservation now runs before the failed-phase early return and resets the stale status.lastScanID binding for terminal (succeeded/failed) mismatched runs as well — only active runs are additionally terminalized. An edited RepositoryScan therefore returns to Pending and starts a current-generation initial run instead of staying Ready against an old-generation run or reusing its incremental baseline.

**Comment by sozercan** [comment]
Closing this PR after extracting the run-lifecycle fixes into #501. That replacement binds runs to RepositoryScan UID/generation, retires stale reservations, prevents old tasks and status writers from taking over a newer run, and uses deterministic stage names per run.

Immutable finding history remains a separate follow-up to scope against the current terminal-result flow. The remaining history, receipt, bundle, and enforcement work in this PR has not been merged or fully superseded by #501.

#501 is still open. Keeping the `agent/security-scan-integrity` branch as a reference for the deferred work.
